clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,740 @@
+  ///                 #             |      # Name_1m
+  ///                 #             |      # ValAddr_1m
+  ///                 #             |      # Type_1m
+  ///                 #             |      # ValSize_1m
+  ///                 #             +-----># ...
+  ///                 #                    #
+  /// \endcode
+  ///
+  /// \returns Pair of pointers to the beginning and end of the property set
+  /// array, or a pair of nullptrs in case the properties file wasn't specified.
+  std::pair<Constant *, Constant *>
+  addPropertySetRegistry(const PropertySetRegistry &PropRegistry) {
+    // transform all property sets to IR and get the middle column image into
+    // the PropSetsInits
+    SmallVector<Constant *> PropSetsInits;
+    for (const auto &PropSet : PropRegistry) {
+      // create content in the rightmost column and get begin/end pointers
+      std::pair<Constant *, Constant *> Props =
+          addPropertySetToModule(PropSet.second);
+      // get the next the middle column element
+      auto *Category = addStringToModule(PropSet.first, "SYCL_PropSetName");
+      PropSetsInits.push_back(ConstantStruct::get(SyclPropSetTy, Category,
+                                                  Props.first, Props.second));
+    }
+    // now get content for the leftmost column - create the top-level
+    // PropertySetsBegin/PropertySetsBegin entries and return pointers to them
+    return addStructArrayToModule(PropSetsInits, SyclPropSetTy);
+  }
+
+  /// Emits a global array that contains \p Address and \P Size. Also add
+  /// it into llvm.used to force it to be emitted in the object file.
+  void emitRegistrationFunctions(Constant *Address, size_t Size, Twine ImageID,
+                                 StringRef OffloadKindTag) {
+    Type *IntPtrTy = M.getDataLayout().getIntPtrType(C);
+    auto *ImgInfoArr =
+        ConstantArray::get(ArrayType::get(IntPtrTy, 2),
+                           {ConstantExpr::getPointerCast(Address, IntPtrTy),
+                            ConstantInt::get(IntPtrTy, Size)});
+    auto *ImgInfoVar = new GlobalVariable(
+        M, ImgInfoArr->getType(), true, GlobalVariable::InternalLinkage,
+        ImgInfoArr, Twine(OffloadKindTag) + ImageID + ".info");
+    ImgInfoVar->setAlignment(
+        MaybeAlign(M.getDataLayout().getTypeStoreSize(IntPtrTy) * 2u));
+    ImgInfoVar->setUnnamedAddr(GlobalValue::UnnamedAddr::Local);
+    ImgInfoVar->setSection(".tgtimg");
+
+    // Add image info to the used list to force it to be emitted to the
+    // object.
+    appendToUsed(M, ImgInfoVar);
+  }
+
+  Constant *wrapImage(const SYCLImage &Image, Twine ImageID,
+                      StringRef OffloadKindTag) {
+    auto *NullPtr = Constant::getNullValue(PointerType::getUnqual(C));
+    // DeviceImageStructVersion change log:
+    // -- version 2: updated to PI 1.2 binary image format
+    constexpr uint16_t DeviceImageStructVersion = 2;
+    constexpr uint8_t SYCLOffloadKind = 4;
```

Comment:
Add comment for why we are setting this to 4?

Thanks

Reply:
added.